sklearn/utils/extmath.py

```diff
@@ -374,6 +374,126 @@ def randomized_svd(M, n_components, *, n_oversamples=10, n_iter='auto',
+        the data, i.e. getting the random vectors to initialize the algorithm.
+        Pass an int for reproducible results across multiple function calls.
+        See :term:`Glossary <random_state>`.
+
+    Notes
+    -----
+    This algorithm finds a (usually very good) approximate truncated
+    eigendecomposition using randomized methods to speed up the computations.
+
+    Strategy 'value': not implemented yet. An algorithm can be found in the
+    Halko et al reference below, for future implementation.
+
+    Strategy 'module':
+    The principle is that for diagonalizable matrices, the singular values and
+    eigenvalues are related: if t is an eigenvalue of A, then :math:`|t|` is a
+    singular value of A. This method relies on a randomized SVD to find the n
+    singular components corresponding to the n singular values with largest
+    modules, and then uses the signs of the singular vectors to find the true
+    sign of t: if the sign of left and right singular vectors are different
+    then the corresponding eigenvalue is negative.
+
+    This method is particularly fast on large matrices on which
+    you wish to extract only a small number of components. In order to
+    obtain further speed up, `n_iter` can be set <=2 (at the cost of
+    loss of precision).
+
+    See Also
+    --------
+    :func:`randomized_svd`
+
+    References
+    ----------
+    * Finding structure with randomness: Stochastic algorithms for constructing
+      approximate matrix decompositions (Algorithm 4.3 for strategy 'module')
+      Halko, et al., 2009 https://arxiv.org/abs/0909.4061
+
+    """
+    if selection == 'value':
+        # to do : an algorithm can be found in the Halko et al reference
+        raise NotImplementedError()
+
+    elif selection == 'module':
+        # Note: no need for deterministic U and Vt (flip_sign=True),
+        # as we only use the dot product UVt afterwards
+        U, S, Vt = randomized_svd(
+            M, n_components=n_components, n_oversamples=n_oversamples,
+            n_iter=n_iter,
+            power_iteration_normalizer=power_iteration_normalizer,
+            flip_sign=False, random_state=random_state)
+
+        alphas_ = U[:, :n_components]  # eigenvectors
+        lambdas_ = S[:n_components]  # eigenvalues
+
+        # Conversion of Singular values into Eigenvalues:
+        # For any eigenvalue t, the corresponding singular value is |t|.
+        # So if there is a negative eigenvalue t, the corresponding singular
+        # value will be -t, and the left (U) and right (V) singular vectors
+        # will have opposite signs.
+        # Fastest way: see <https://stackoverflow.com/a/61974002/7262247>
+        diag_VtU = np.einsum('ji,ij->j',
+                             Vt[:n_components, :], U[:, :n_components])
+        signs = np.sign(diag_VtU)
+        lambdas_ = lambdas_ * signs
+
+    else:
+        raise ValueError("Invalid `selection`: %r" % selection)
```

Comment:
I would prefer to do input validation in the beginning:

```py
if selection != 'module':
	raise ValueError(...)
```

Reply:
I cannot, actually: I need to raise a specific error for `selection == 'value'`

Reply:
I see now. It is a little strange that the default value of `selection='value'` is the default and not implemented.

Does it make sense to remove the `selection='value'` and have the default be `selection='module'` for now? This way we can remove the `selection='value'` keyword argument.

Reply:
In general, the best and most natural way to solve the `eigh` problem with the randomized method is `selection='value'`. This is why it is the default method, in order to match the API from all other `xxx_eigh` methods. That would be strange to change the API default value across time, so this is to prepare for the time when it is implemented.

But this PR is about KernelPCA, where the Gram matrix is supposed to be positive semi-definite (PSD). For this reason the other method `selection='module'` is also ok. And it has the advantage of relying on `randomized_svd` which is already implemented and validated in scikit-learn. So this is why I suggest to merge this PR with this already proven method, and then in a later step augment `randomized_eigh` with the other method so that it can for example allow the randomized method to be used as the default in isomap (where the matrix is not necessary PSD). My plan is to open an issue to fill this section as soon as this PR is merged.

Reply:
I see your line of reasoning.

> That would be strange to change the API default value across time, so this is to prepare for the time when it is implemented.

Can we make `randomized_eigh` private for now, i.e., `utils.extmath._randomized_eigh`? This way we do not expose a public function where the default is not implemented. When `selection='value'` is implemented, then we can make the function public.

Reply:
I'm ok with this if you think that it's better. I'll do this now